### PR TITLE
feat(portal): 139 create server side portal

### PR DESCRIPTION
**Pull request by Tzal3x**
Create a nextjs server that acts as a server-side walrus-sites portal.
It works by combining a middleware for preprocessing the requests, and a root endpoint (defined in`route.ts`) that actually contains the logic as the service worker did. 

The middleware has the following role:
1. To handle requests like `http://<name>.walrus.site/path/to/file/` without throwing an error. Explanation: if you don't define a route on nextjs, the default logic is to throw an error because the `/path/to/file` endpoint does not exist. 
2. To delegate the whole request (including the subnames) to the route.ts in order to fetch the site resources. By default the subnames are removed from the hostname in nextjs. 

In summary, the middleware is just a delegator and an error-preventer of the original url. 

💁‍♂️ **Review tip**: the PR consists of 2 commits. 
I would suggest reviewing commit by commit and not the whole PR altogether.

### CODE REVIEW

**Comment by Tzal3x** [comment]
Base skeleton is there. It doesn't work yet, something wrong happens with the url parsing.

**Comment by Tzal3x** [comment]
To check: https://github.com/dcangulo/nextjs-subdomain-example

**Comment by Tzal3x** [comment]
Another article to consider: [nextjs wildcard subdomains](https://www.smashingmagazine.com/2021/11/nextjs-wildcard-subdomains/)

**Comment by Tzal3x** [comment]
Looks like subdomains don't work with safari. 

i.e. http://docs.localhost:3000/ returns this error:
`Safari can't find server. Safari can't find the server docs.localhost`

File: portal/server/.prettierrc
Comment (giac-mysten) [inline portal/server/.prettierrc:1]: Suggestion: I would prefer this to be integrated with `.editorconfig` . Afaics Prettier does parse the editorconfig and uses it for formatting.

File: portal/server/app/route.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import { getDomain, getSubdomainAndPath } from '@lib/domain_parsing'
+import { redirectToAggregatorUrlResponse, redirectToPortalURLResponse } from '@lib/redirects'
+import { getBlobIdLink, getObjectIdLink } from '@lib/links'
+import { resolveAndFetchPage } from '@lib/page_fetching'
+
+export async function GET(req: Request) {
+    const originalUrl = req.headers.get('x-original-url')
+    if (!originalUrl) {
+        throw new Error('No original url found in request headers')
+    }
```
Comment (giac-mysten) [inline portal/server/app/route.ts:13]: Question: why do we need this? do we expect proxies changing the request?

File: portal/server/custom.d.ts
Comment (giac-mysten) [inline portal/server/custom.d.ts:1]: Question: I'm guessing there was no way to import this definition from from `common/`? (probably not a big deal to have this file duplicated)
(edit: apparently there is a way, by exporting this as type definitions in the `package.json`)

File: portal/server/middleware.ts
Comment (giac-mysten) [inline portal/server/middleware.ts:1]: Question: do we still need the middleware?

**Review by giac-mysten — COMMENTED** [review]
great stuff @Tzal3x . I made a first pass on the PR. We can address the problems you mention above offline.

File: portal/server/middleware.ts
Comment (Tzal3x) [reply portal/server/middleware.ts:1]: Yes because it needs to delegate the original url. If we don't use it, nextjs removes any subnames, and also it will throw an error if you try to access a path (endpoint) that does not exist like `docs.portal/path/to/something`

File: portal/server/app/route.ts (hunk 1 above)
Comment (Tzal3x) [reply portal/server/app/route.ts:13]: `x-original-url` is a custom header that contains the original url that was requested by the browser. 
I had to use it because there was no other way accessing the subdomains.

File: portal/server/.prettierrc
Comment (Tzal3x) [reply portal/server/.prettierrc:1]: You are right, a lot of things contained in prettier are also used in .editorconfig. 
That will be an easy fix.

File: portal/server/custom.d.ts
Comment (Tzal3x) [reply portal/server/custom.d.ts:1]: Cool, I managed to make it work using a [types](https://www.typescriptlang.org/tsconfig/#types) declaration in `server/tsconfig.json`. 

For some reason it doesn't work by defining this in `package.json`, even though we are given that option. 🤷

File: portal/server/.prettierrc
Comment (giac-mysten) [reply portal/server/.prettierrc:1]: did you add the settings to `.editorconfig`? if so, can you commit them?

**Review by giac-mysten — APPROVED** [review]
Thanks! Let's merge it as is, and investigate further if we can remove the middleware (#142 )

**Comment by Tzal3x** [comment]
> did you add the settings to .editorconfig? if so, can you commit them?

@giac-mysten I actually juste removed the prettierrc file because it was redundant after I compared them.
